packages/contracts/contracts/L1/ValidatorManager.sol

```diff
@@ -0,0 +1,637 @@
+
+    /**
+     * @notice A mapping of the validator to the validator information.
+     */
+    mapping(address => Validator) internal _validatorInfo;
+
+    /**
+     * @notice A mapping of the jailed validator to the jail expiration timestamp.
+     */
+    mapping(address => uint128) internal _jail;
+
+    /**
+     * @notice A mapping of output index challenged successfully to pending challenge rewards.
+     */
+    mapping(uint256 => uint128) internal _pendingChallengeReward;
+
+    /**
+     * @notice A modifier that only allows L2OutputOracle contract to call.
+     */
+    modifier onlyL2OutputOracle() {
+        if (msg.sender != address(L2_ORACLE)) revert NotAllowedCaller();
+        _;
+    }
+
+    /**
+     * @notice A modifier that only allows Colosseum contract to call.
+     */
+    modifier onlyColosseum() {
+        if (msg.sender != L2_ORACLE.COLOSSEUM()) revert NotAllowedCaller();
+        _;
+    }
+
+    /**
+     * @notice Semantic version.
+     * @custom:semver 1.0.0
+     */
+    string public constant version = "1.0.0";
+
+    /**
+     * @notice Constructs the ValidatorManager contract.
+     *
+     * @param _constructorParams The constructor parameters.
+     */
+    constructor(ConstructorParams memory _constructorParams) {
+        if (_constructorParams._minRegisterAmount > _constructorParams._minStartAmount)
+            revert InvalidConstructorParams();
+
+        L2_ORACLE = _constructorParams._l2Oracle;
+        ASSET_MANAGER = _constructorParams._assetManager;
+        TRUSTED_VALIDATOR = _constructorParams._trustedValidator;
+        MIN_REGISTER_AMOUNT = _constructorParams._minRegisterAmount;
+        MIN_START_AMOUNT = _constructorParams._minStartAmount;
+        COMMISSION_RATE_MIN_CHANGE_SECONDS = _constructorParams._commissionRateMinChangeSeconds;
+        // Note that this value MUST be (SUBMISSION_INTERVAL * L2_BLOCK_TIME) / 2.
+        ROUND_DURATION_SECONDS = _constructorParams._roundDurationSeconds;
+        JAIL_PERIOD_SECONDS = _constructorParams._jailPeriodSeconds;
+        JAIL_THRESHOLD = _constructorParams._jailThreshold;
+        MAX_OUTPUT_FINALIZATIONS = _constructorParams._maxOutputFinalizations;
+        BASE_REWARD = _constructorParams._baseReward;
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function registerValidator(
+        uint128 assets,
+        uint8 commissionRate,
+        uint8 commissionMaxChangeRate
+    ) external {
+        if (getStatus(msg.sender) != ValidatorStatus.NONE) revert ImproperValidatorStatus();
+
+        if (assets < MIN_REGISTER_AMOUNT) revert InsufficientAsset();
+
+        if (commissionRate > COMMISSION_RATE_DENOM) revert MaxCommissionRateExceeded();
+
+        if (commissionMaxChangeRate > COMMISSION_RATE_DENOM)
+            revert MaxCommissionChangeRateExceeded();
+
+        Validator storage validatorInfo = _validatorInfo[msg.sender];
+        validatorInfo.isInitiated = true;
```

Comment:
There is no way to check if the validator is registered or not using `validatorInfo`. Because except for `isInitiated`, the other fields can be 0 even in the case validator has registered.